provider/slof.py

```diff
@@ -0,0 +1,142 @@
+        test.error('No found serial log in %s sec.' % file_timeout)
+
+    while time.time() < (float(timeout) + time.time()):
+        content, cur_pos = get_boot_content(vm, start_pos, start_str, end_str)
+        if content:
+            if debug:
+                logging.info('Output of SLOF:\n%s' % ''.join(content))
+            return content, cur_pos
+    test.fail('No found corresponding SLOF info in serial log during %s sec.' %
+              timeout)
+
+
+def get_boot_device(content):
+    """
+    Get the device info which tried to load.
+    :param content: SLOF content.
+    :return: device booted.
+    """
+
+    for line in content:
+        device = re.search(r'(\s+Trying to load:\s+from:\s)(/.+)(\s+\.\.\.)',
+                           line)
+        if device:
+            yield device.group(2)
+
+
+def verify_boot_device(content, parent_bus_type, child_bus_type, child_addr,
+                       sub_child_addr=None):
+    """
+    Verify guest whether booted with specified device.
+    :param content: SLOF content list.
+    :param parent_bus_type: str, the type of parent bus of device.
+    :param child_bus_type: str, the type of bus of device.
+    :param child_addr: str, the address of device bus.
+    :param sub_child_addr: str, the address of device child bus.
+    :return: True if booted from the specified device.
+    """
+
+    if '0x' in child_addr:
+        addr = child_addr[2:].lstrip('0')
+    else:
+        addr = child_addr
+    if sub_child_addr:
+        if '0x' in sub_child_addr:
+            sub_addr = sub_child_addr[2:].lstrip('0')
+        else:
+            sub_addr = sub_child_addr
+
+    for index, line in enumerate(content):
+        if 'Trying to load' in line:
+            logging.info('Booted device info:\n%s' % ''.join(content[index:]))
+            break
+
+    pattern = re.compile(r'/\w+.{1}\w+@')
+    for name in get_boot_device(content):
+        if parent_bus_type == 'pci':
+            # virtio-blk, virtio-scsi and ethernet device.
+            if child_bus_type == 'scsi' or child_bus_type == 'ethernet':
+                if addr == pattern.split(name)[2]:
+                    return True
+            # pci-bridge, usb device.
+            elif child_bus_type == 'pci-bridge' or child_bus_type == 'usb':
+                if (addr == pattern.split(name)[2] and
+                        sub_addr == pattern.split(name)[3]):
+                    return True
+        elif parent_bus_type == 'vdevice':
+            if child_bus_type == 'v-scsi':
+                if addr == pattern.split(name)[1]:
+                    return True
+    return False
+
+
+def check_error(test, content):
+    """
+    Check if there are error info in SLOF content.
+    :param test: kvm test object.
+    :param content: SLOF content list.
+    """
+    for _, line in enumerate(content):
+        if re.search(r'error', line, re.IGNORECASE):
+            test.fail('Found errors: %s' % line)
+
```

Comment:
redundant line here.

Reply:
We should consider another situation that slof reboot or there is duplicate content in slof output, similar to the following bug:
Bug 1443904 - SLOF user interface display bootable device and "SLOF" words duplicate
(https://bugzilla.redhat.com/show_bug.cgi?id=1443904)

Reply:
As we talked, add two cases for covering them.